src/Controls/src/Core/Binding.cs

```diff
@@ -120,7 +122,13 @@ internal override void Apply(object context, BindableObject bindObj, BindablePro
 			object src = _source;
 			var isApplied = IsApplied;
 
-			base.Apply(src ?? context, bindObj, targetProperty, fromBindingContextChanged, specificity);
+			var bindingContext = src ?? Context ?? context;
+			if (DataType != null && bindingContext != null && !DataType.IsAssignableFrom(bindingContext.GetType()))
+			{
+				bindingContext = null;
```

Comment:
Should we log a warning to make it easier to debug the type mismatch?

Reply:
I'm afraid that a warning will make HotReload fail

Reply:
If it's something that "works" but is not the correct or preferred way to bind, I rather hot reload not work and tell the user about the issue rather than let it through even though there is a type mismatch that "works" but is wrong. I'm also not sure if warnings in it of themselves would stop hot reload itself or if we could still show the warning and let the action through anyway.

@mgoertz-msft what do you think?

Reply:
You would need to ask @etvorun. I'm not sure what the exact condition is for errors/warnings to cause Hot Reload to skip requests. I also thought that HR only looks at these reported from IntelliSense and not build but I could be wrong.

Reply:
My preference would be to follow WPF, which is to silently ignore bad binding and fire [BindingErrorEventArgs](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.xaml.diagnostics.bindingerroreventargs?view=net-maui-8.0). Then extend VS support for [binding diagnostics](https://learn.microsoft.com/en-us/visualstudio/xaml-tools/xaml-data-binding-diagnostics?view=vs-2022) for MAUI. @spadapet is the expert in that area.

Reply:
I agree with @etvorun. XAML Hot Reload will send the bad binding to the app since XAML parsing is valid (like @mgoertz-msft said). If this code triggers a [BindingFailed](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.xaml.diagnostics.bindingdiagnostics.bindingfailed?view=net-maui-8.0) event, then the UI in Visual Studio and the in-app toolbar will immediately show that a runtime binding failure occurred. Then the user can see a detailed error message and jump the XAML.

![image](https://github.com/dotnet/maui/assets/2523431/4dafd1ef-a403-4e6f-a3a6-4bd4e133cf35)

Reply:
I'm now sending a binding failure in case of mismatch. Recall that this could be transient (as the context might not be set until the parent is), and a transient error shouldn't make HR to fail